### ISSUE TITLE
Add missing test modules in our unit test suite

### ISSUE THREAD

**Issue by potiuk** [apache/airflow#35442]
See `test_providers_modules_should_have_tests` in  `tests/always/test_project_structure.py` but those tests are missing currently:

## Provider [amazon](https://github.com/apache/airflow/tree/main/tests/providers/amazon)
[codecov](https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders%2Famazon)
- [ ] tests/providers/amazon/aws/executors/ecs/test_boto_schema.py
- [ ] tests/providers/amazon/aws/executors/ecs/test_ecs_executor_config.py
- [ ] tests/providers/amazon/aws/executors/ecs/test_utils.py
- [x] tests/providers/amazon/aws/fs/test_s3.py
- [x] tests/providers/amazon/aws/hooks/test_dms.py
- [x] tests/providers/amazon/aws/links/test_base_aws.py
- [x] tests/providers/amazon/aws/links/test_batch.py
- [x] tests/providers/amazon/aws/links/test_emr.py
- [x] tests/providers/amazon/aws/links/test_glue.py
- [x] tests/providers/amazon/aws/links/test_logs.py
- [x] tests/providers/amazon/aws/operators/test_dms.py
- [ ] tests/providers/amazon/aws/operators/test_emr.py
- [x] tests/providers/amazon/aws/operators/test_s3.py
- [ ] tests/providers/amazon/aws/operators/test_sagemaker.py
- [x] tests/providers/amazon/aws/sensors/test_dms.py
- [ ] tests/providers/amazon/aws/sensors/test_emr.py
- [x] tests/providers/amazon/aws/sensors/test_s3.py
- [ ] tests/providers/amazon/aws/sensors/test_sagemaker.py
- [ ] tests/providers/amazon/aws/test_exceptions.py
- [x] tests/providers/amazon/aws/triggers/test_athena.py
- [x] tests/providers/amazon/aws/triggers/test_batch.py
- [ ] tests/providers/amazon/aws/triggers/test_eks.py
- [x] tests/providers/amazon/aws/triggers/test_emr.py
- [x] tests/providers/amazon/aws/triggers/test_glue_crawler.py
- [x] tests/providers/amazon/aws/triggers/test_lambda_function.py
- [ ] tests/providers/amazon/aws/triggers/test_rds.py
- [x] tests/providers/amazon/aws/triggers/test_redshift_cluster.py
- [ ] tests/providers/amazon/aws/triggers/test_step_function.py
- [ ] tests/providers/amazon/aws/utils/test_rds.py
- [ ] tests/providers/amazon/aws/utils/test_sagemaker.py
- [x] tests/providers/amazon/aws/utils/test_sqs.py
- [x] tests/providers/amazon/aws/utils/test_tags.py
- [ ] tests/providers/amazon/aws/waiters/test_base_waiter.py

## Provider [apache.cassandra](https://github.com/apache/airflow/tree/main/tests/providers/apache/cassandra)
[codecov](https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders%2Fapache%2Fcassandra)
- [ ] tests/providers/apache/cassandra/hooks/test_cassandra.py

## Provider [apache.drill](https://github.com/apache/airflow/tree/main/tests/providers/apache/drill)
[codecov](https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders%2Fapache%2Fdrill)
- [ ] tests/providers/apache/drill/operators/test_drill.py

## Provider [apache.druid](https://github.com/apache/airflow/tree/main/tests/providers/apache/druid)
[codecov](https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders%2Fapache%2Fdruid)
- [ ] tests/providers/apache/druid/operators/test_druid_check.py

## Provider [apache.hdfs](https://github.com/apache/airflow/tree/main/tests/providers/apache/hdfs)
[codecov](https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders%2Fapache%2Fhdfs)
- [x] tests/providers/apache/hdfs/hooks/test_hdfs.py
- [ ] tests/providers/apache/hdfs/log/test_hdfs_task_handler.py
- [x] tests/providers/apache/hdfs/sensors/test_hdfs.py

## Provider [apache.hive](https://github.com/apache/airflow/tree/main/tests/providers/apache/hive)
[codecov](https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders%2Fapache%2Fhive)
- [ ] tests/providers/apache/hive/plugins/test_hive.py
- [ ] tests/providers/apache/hive/transfers/test_mssql_to_hive.py

## Provider [apache.kafka](https://github.com/apache/airflow/tree/main/tests/providers/apache/kafka)
[codecov](https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders%2Fapache%2Fkafka)
- [ ] tests/providers/apache/kafka/hooks/test_base.py
- [x] tests/providers/apache/kafka/hooks/test_client.py
- [x] tests/providers/apache/kafka/hooks/test_consume.py
- [x] tests/providers/apache/kafka/hooks/test_produce.py

## Provider [celery](https://github.com/apache/airflow/tree/main/tests/providers/celery)
[codecov](https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders%2Fcelery)
- [ ] tests/providers/celery/executors/test_celery_executor_utils.py
- [ ] tests/providers/celery/executors/test_default_celery.py

## Provider [cncf.kubernetes](https://github.com/apache/airflow/tree/main/tests/providers/cncf/kubernetes)
[codecov](https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders%2Fcncf%2Fkubernetes)
- [ ] tests/providers/cncf/kubernetes/backcompat/test_backwards_compat_converters.py
- [ ] tests/providers/cncf/kubernetes/executors/test_kubernetes_executor_types.py
- [ ] tests/providers/cncf/kubernetes/executors/test_kubernetes_executor_utils.py
- [ ] tests/providers/cncf/kubernetes/operators/test_kubernetes_pod.py
- [ ] tests/providers/cncf/kubernetes/test_k8s_model.py
- [ ] tests/providers/cncf/kubernetes/test_kube_client.py
- [ ] tests/providers/cncf/kubernetes/test_kube_config.py
- [ ] tests/providers/cncf/kubernetes/test_pod_generator_deprecated.py
- [ ] tests/providers/cncf/kubernetes/test_pod_launcher_deprecated.py
- [ ] tests/providers/cncf/kubernetes/test_python_kubernetes_script.py
- [ ] tests/providers/cncf/kubernetes/test_secret.py
- [ ] tests/providers/cncf/kubernetes/triggers/test_kubernetes_pod.py
- [ ] tests/providers/cncf/kubernetes/utils/test_delete_from.py
- [ ] tests/providers/cncf/kubernetes/utils/test_k8s_hashlib_wrapper.py
- [ ] tests/providers/cncf/kubernetes/utils/test_xcom_sidecar.py

## Provider [common.io](https://github.com/apache/airflow/tree/main/tests/providers/common/io)
[codecov](https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders%2Fcommon%2Fio)
- [x] tests/providers/common/io/operators/test_file_transfer.py

## Provider [databricks](https://github.com/apache/airflow/tree/main/tests/providers/databricks)
[codecov](https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders%2Fdatabricks)
- [x] tests/providers/databricks/hooks/test_databricks_base.py

## Provider [dbt.cloud](https://github.com/apache/airflow/tree/main/tests/providers/dbt/cloud)
[codecov](https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders%2Fdbt%2Fcloud)
- [x] tests/providers/dbt/cloud/hooks/test_dbt.py
- [x] tests/providers/dbt/cloud/operators/test_dbt.py
- [x] tests/providers/dbt/cloud/sensors/test_dbt.py
- [x] tests/providers/dbt/cloud/triggers/test_dbt.py

## Provider [docker](https://github.com/apache/airflow/tree/main/tests/providers/docker)
[codecov](https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders%2Fdocker)
- [ ] tests/providers/docker/test_exceptions.py

## Provider [elasticsearch](https://github.com/apache/airflow/tree/main/tests/providers/elasticsearch)
[codecov](https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders%2Felasticsearch)
- [x] tests/providers/elasticsearch/log/test_es_json_formatter.py
- [x] tests/providers/elasticsearch/log/test_es_response.py

## Provider [google](https://github.com/apache/airflow/tree/main/tests/providers/google)
[codecov](https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders%2Fgoogle)
- [ ] tests/providers/google/cloud/fs/test_gcs.py
- [ ] tests/providers/google/cloud/links/test_automl.py
- [x] tests/providers/google/cloud/links/test_base.py
- [x] tests/providers/google/cloud/links/test_bigquery.py
- [ ] tests/providers/google/cloud/links/test_bigquery_dts.py
- [ ] tests/providers/google/cloud/links/test_bigtable.py
- [ ] tests/providers/google/cloud/links/test_cloud_build.py
- [ ] tests/providers/google/cloud/links/test_cloud_functions.py
- [ ] tests/providers/google/cloud/links/test_cloud_memorystore.py
- [ ] tests/providers/google/cloud/links/test_cloud_sql.py
- [ ] tests/providers/google/cloud/links/test_cloud_storage_transfer.py
- [ ] tests/providers/google/cloud/links/test_cloud_tasks.py
- [ ] tests/providers/google/cloud/links/test_compute.py
- [ ] tests/providers/google/cloud/links/test_data_loss_prevention.py
- [ ] tests/providers/google/cloud/links/test_datacatalog.py
- [ ] tests/providers/google/cloud/links/test_dataflow.py
- [ ] tests/providers/google/cloud/links/test_dataform.py
- [ ] tests/providers/google/cloud/links/test_datafusion.py
- [ ] tests/providers/google/cloud/links/test_dataplex.py
- [ ] tests/providers/google/cloud/links/test_dataprep.py
- [ ] tests/providers/google/cloud/links/test_dataproc.py
- [ ] tests/providers/google/cloud/links/test_datastore.py
- [ ] tests/providers/google/cloud/links/test_kubernetes_engine.py
- [ ] tests/providers/google/cloud/links/test_life_sciences.py
- [ ] tests/providers/google/cloud/links/test_mlengine.py
- [ ] tests/providers/google/cloud/links/test_pubsub.py
- [ ] tests/providers/google/cloud/links/test_spanner.py
- [ ] tests/providers/google/cloud/links/test_stackdriver.py
- [ ] tests/providers/google/cloud/links/test_vertex_ai.py
- [ ] tests/providers/google/cloud/links/test_workflows.py
- [ ] tests/providers/google/cloud/operators/vertex_ai/test_auto_ml.py
- [ ] tests/providers/google/cloud/operators/vertex_ai/test_batch_prediction_job.py
- [ ] tests/providers/google/cloud/operators/vertex_ai/test_custom_job.py
- [ ] tests/providers/google/cloud/operators/vertex_ai/test_dataset.py
- [ ] tests/providers/google/cloud/operators/vertex_ai/test_endpoint_service.py
- [ ] tests/providers/google/cloud/operators/vertex_ai/test_hyperparameter_tuning_job.py
- [ ] tests/providers/google/cloud/operators/vertex_ai/test_model_service.py
- [ ] tests/providers/google/cloud/operators/vertex_ai/test_pipeline_job.py
- [ ] tests/providers/google/cloud/sensors/test_dataform.py
- [ ] tests/providers/google/cloud/transfers/test_bigquery_to_sql.py
- [ ] tests/providers/google/cloud/transfers/test_presto_to_gcs.py
- [ ] tests/providers/google/cloud/transfers/test_trino_to_gcs.py
- [ ] tests/providers/google/cloud/triggers/test_cloud_composer.py
- [ ] tests/providers/google/cloud/transfers/test_bigquery_to_mssql.py
- [ ] tests/providers/google/cloud/transfers/test_mssql_to_gcs.py
- [ ] tests/providers/google/cloud/utils/test_bigquery.py
- [ ] tests/providers/google/cloud/utils/test_bigquery_get_data.py
- [ ] tests/providers/google/cloud/utils/test_dataform.py
- [ ] tests/providers/google/common/links/test_storage.py
- [ ] tests/providers/google/common/test_consts.py
- [ ] tests/providers/google/test_go_module_utils.py

## Provider [microsoft.azure](https://github.com/apache/airflow/tree/main/tests/providers/microsoft/azure)
[codecov](https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders%2Fmicrosoft%2Fazure)
- [ ] tests/providers/microsoft/azure/fs/test_adls.py
- [x] tests/providers/microsoft/azure/hooks/test_batch.py
- [x] tests/providers/microsoft/azure/hooks/test_container_instance.py
- [x] tests/providers/microsoft/azure/hooks/test_container_registry.py
- [x] tests/providers/microsoft/azure/hooks/test_container_volume.py
- [x] tests/providers/microsoft/azure/hooks/test_cosmos.py
- [x] tests/providers/microsoft/azure/hooks/test_data_factory.py
- [x] tests/providers/microsoft/azure/hooks/test_data_lake.py
- [x] tests/providers/microsoft/azure/hooks/test_fileshare.py
- [x] tests/providers/microsoft/azure/hooks/test_synapse.py
- [ ] tests/providers/microsoft/azure/operators/test_adls.py
- [x] tests/providers/microsoft/azure/operators/test_batch.py
- [x] tests/providers/microsoft/azure/operators/test_container_instances.py
- [x] tests/providers/microsoft/azure/operators/test_cosmos.py
- [x] tests/providers/microsoft/azure/operators/test_data_factory.py
- [x] tests/providers/microsoft/azure/operators/test_synapse.py
- [x] tests/providers/microsoft/azure/secrets/test_key_vault.py
- [x] tests/providers/microsoft/azure/sensors/test_cosmos.py
- [x] tests/providers/microsoft/azure/sensors/test_data_factory.py
- [ ] tests/providers/microsoft/azure/transfers/test_azure_blob_to_gcs.py
- [x] tests/providers/microsoft/azure/triggers/test_data_factory.py
- [ ] tests/providers/microsoft/azure/triggers/test_wasb.py

## Provider [mongo](https://github.com/apache/airflow/tree/main/tests/providers/mongo)
[codecov](https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders%2Fmongo)
- [ ] tests/providers/mongo/sensors/test_mongo.py

## Provider [openlineage](https://github.com/apache/airflow/tree/main/tests/providers/openlineage)
[codecov](https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders%2Fopenlineage)
- [x] tests/providers/openlineage/extractors/test_base.py
- [x] tests/providers/openlineage/extractors/test_bash.py
- [ ] tests/providers/openlineage/extractors/test_manager.py
- [x] tests/providers/openlineage/extractors/test_python.py
- [ ] tests/providers/openlineage/plugins/test_adapter.py
- [ ] tests/providers/openlineage/plugins/test_facets.py
- [ ] tests/providers/openlineage/plugins/test_macros.py
- [ ] tests/providers/openlineage/test_sqlparser.py

## Provider [presto](https://github.com/apache/airflow/tree/main/tests/providers/presto)
[codecov](https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders%2Fpresto)
- [x] tests/providers/presto/transfers/test_gcs_to_presto.py

## Provider [redis](https://github.com/apache/airflow/tree/main/tests/providers/redis)
[codecov](https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders%2Fredis)
- [ ] tests/providers/redis/operators/test_redis_publish.py
- [ ] tests/providers/redis/sensors/test_redis_key.py

## Provider [slack](https://github.com/apache/airflow/tree/main/tests/providers/slack)
[codecov](https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders%2Fslack)
- [ ] tests/providers/slack/notifications/test_slack_notifier.py

## Provider [snowflake](https://github.com/apache/airflow/tree/main/tests/providers/snowflake)
[codecov](https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders%2Fsnowflake)
- [ ] tests/providers/snowflake/triggers/test_snowflake_trigger.py

## Provider [trino](https://github.com/apache/airflow/tree/main/tests/providers/trino)
[codecov](https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders%2Ftrino)
- [x] tests/providers/trino/transfers/test_gcs_to_trino.py

### Committer

- [X] I acknowledge that I am a maintainer/committer of the Apache Airflow project.

**Comment by potiuk** [comment]
cc: @eladkal

**Comment by potiuk** [comment]
@eladkal -> first test removed from the list with #35457 -> I hope you can bring more people to contribute here :)

**Comment by potiuk** [comment]
Also related to #35127

**Comment by potiuk** [comment]
cc: @ephraimbuddy -> this might guide people on where to add tests to cover whole modules that are not "targeted"  for tests (they might still be accidentally tested by other modules but they do not have dedicated module tests) - this might help people to make decision on where to add some tests. Also this is a nice list of "todos"

**Comment by Taragolis** [comment]
> `tests/providers/slack/notifications/test_slack_notifier.py`

I think how to exclude case when the original module `airflow.providers.slack.notifications.slack_notifier` is deprecated

**Comment by Taragolis** [comment]
And just add here previous investigation about missing tests https://github.com/apache/airflow/pull/28459#discussion_r1054708003, non all of them actually missing:

> * some of tests moved into tests.integrations
> * Amazon links tested in one go in same module
> * Some of amazon tests split across different test modules, for example: test_s3_bucket.py + test_s3_bucket_tagging.py + test_s3_file_transform.py + test_s3_list.py + test_s3_list_prefixes.py + test_s3_object.py but module named as s3.py (merged version)

**Comment by Taragolis** [comment]
And `codecov` statistic also might help to found untested parts: https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders

**Comment by eladkal** [comment]
> > `tests/providers/slack/notifications/test_slack_notifier.py`
> 
> I think how to exclude case when the original module `airflow.providers.slack.notifications.slack_notifier` is deprecated

In similar cases we used to have deprecated classes list and we removed it from the output

**Comment by potiuk** [comment]
> I think how to exclude case when the original module airflow.providers.slack.notifications.slack_notifier is deprecated

Absolutely - I think there are many false-negatives there, so removal them by finding and automated way and improving the test case is absolutely legitimate way of cleaning some of those. PRs are most welcome :). I just *revived* the test  really quickly - also in order to not allow more of new ones creep in, but I think we should definitely remove some of those from the list by making the test "smarter"

> And codecov statistic also might help to found untested parts: https://app.codecov.io/gh/apache/airflow/tree/main/airflow%2Fproviders

Oh absolutely - this is why I added my comment https://github.com/apache/airflow/issues/35442#issuecomment-1793747405 and updated https://github.com/apache/airflow/issues/35127 with the comment that these two are connected. I see that as two sides of the same coing - eventually they should converge - the missing coverage should quite closely show similar list as the one generated here.  Currently they are far-off - for the reasons you explained, but also because some of the module's code are tested by some other tests. - but this is precisely I think what should guide us in making the test smarter - eventually they should show the same result.

> And just add here previous investigation about missing tests https://github.com/apache/airflow/pull/28459#discussion_r1054708003, non all of them actually missing:

> some of tests moved into tests.integrations
> Amazon links tested in one go in same module
> Some of amazon tests split across different test modules, for example: test_s3_bucket.py + test_s3_bucket_tagging.py + test_s3_file_transform.py + test_s3_list.py + test_s3_list_prefixes.py + test_s3_object.py but module named as s3.py (merged version)

Yeah - why don't we automate that in the test then? I think some of them can be easily automatically detected.

**Comment by Taragolis** [comment]
If you don't mind I've split this list by providers

**Comment by potiuk** [comment]
> If you don't mind I've split this list by providers

Not at all ! Very helpful I think :)

**Comment by potiuk** [comment]
Super useful with codecov links BTW! Good idea.

**Comment by mateuslatrova** [comment]
Hi! I found that the classes present in the following modules are already covered with tests, but they are unchecked in the issue's first comment:

tests/providers/amazon/aws/operators/test_emr.py
tests/providers/amazon/aws/operators/test_sagemaker.py
tests/providers/amazon/aws/sensors/test_emr.py
tests/providers/amazon/aws/triggers/test_athena.py
tests/providers/amazon/aws/triggers/test_batch.py
tests/providers/amazon/aws/triggers/test_emr.py
tests/providers/amazon/aws/triggers/test_glue_crawler.py
tests/providers/amazon/aws/triggers/test_lambda_function.py
tests/providers/amazon/aws/triggers/test_rds.py
tests/providers/amazon/aws/triggers/test_redshift_cluster.py
tests/providers/amazon/aws/utils/test_rds.py
tests/providers/amazon/aws/utils/test_sagemaker.py
tests/providers/amazon/aws/utils/test_sqs.py
tests/providers/amazon/aws/utils/test_tags.py
tests/providers/amazon/aws/waiters/test_base_waiter.py
tests/providers/amazon/aws/triggers/test_step_function.py

If it is possible, it would be nice if anyone could check them, so that who is looking to contribute can find missing tests more easily. Thanks!

CC: @potiuk @Taragolis

**Comment by Taragolis** [comment]
@mateuslatrova Some test modules from your list do not exists or named differently

### Exists

I've marked this as completed

 - [tests/providers/amazon/aws/triggers/test_athena.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/triggers/test_athena.py)
 - [tests/providers/amazon/aws/triggers/test_batch.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/triggers/test_batch.py)
 - [tests/providers/amazon/aws/triggers/test_emr.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/triggers/test_emr.py)
 - [tests/providers/amazon/aws/triggers/test_glue_crawler.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/triggers/test_glue_crawler.py)
 - [tests/providers/amazon/aws/triggers/test_lambda_function.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/triggers/test_lambda_function.py)
 - [tests/providers/amazon/aws/triggers/test_rds.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/triggers/test_rds.py)
 - [tests/providers/amazon/aws/triggers/test_redshift_cluster.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/triggers/test_redshift_cluster.py)
 - [tests/providers/amazon/aws/utils/test_sqs.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/utils/test_sqs.py)
 - [tests/providers/amazon/aws/utils/test_tags.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/utils/test_tags.py)
 

### Not Exists
 - [tests/providers/amazon/aws/operators/test_emr.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/operators/test_emr.py)
 - [tests/providers/amazon/aws/operators/test_sagemaker.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/operators/test_sagemaker.py)
 - [tests/providers/amazon/aws/sensors/test_emr.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/sensors/test_emr.py)
 - [tests/providers/amazon/aws/utils/test_rds.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/utils/test_rds.py)
 - [tests/providers/amazon/aws/utils/test_sagemaker.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/utils/test_sagemaker.py)
 - [tests/providers/amazon/aws/waiters/test_base_waiter.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/waiters/test_base_waiter.py)
 - [tests/providers/amazon/aws/triggers/test_step_function.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/triggers/test_step_function.py)

**Comment by mateuslatrova** [comment]
> @mateuslatrova Some test modules from your list do not exists or named differently
> 
> ### Exists
> I've marked this as completed
> 
> * [tests/providers/amazon/aws/triggers/test_athena.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/triggers/test_athena.py)
> * [tests/providers/amazon/aws/triggers/test_batch.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/triggers/test_batch.py)
> * [tests/providers/amazon/aws/triggers/test_emr.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/triggers/test_emr.py)
> * [tests/providers/amazon/aws/triggers/test_glue_crawler.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/triggers/test_glue_crawler.py)
> * [tests/providers/amazon/aws/triggers/test_lambda_function.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/triggers/test_lambda_function.py)
> * [tests/providers/amazon/aws/triggers/test_rds.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/triggers/test_rds.py)
> * [tests/providers/amazon/aws/triggers/test_redshift_cluster.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/triggers/test_redshift_cluster.py)
> * [tests/providers/amazon/aws/utils/test_sqs.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/utils/test_sqs.py)
> * [tests/providers/amazon/aws/utils/test_tags.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/utils/test_tags.py)
> 
> ### Not Exists
> * [tests/providers/amazon/aws/operators/test_emr.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/operators/test_emr.py)
> * [tests/providers/amazon/aws/operators/test_sagemaker.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/operators/test_sagemaker.py)
> * [tests/providers/amazon/aws/sensors/test_emr.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/sensors/test_emr.py)
> * [tests/providers/amazon/aws/utils/test_rds.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/utils/test_rds.py)
> * [tests/providers/amazon/aws/utils/test_sagemaker.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/utils/test_sagemaker.py)
> * [tests/providers/amazon/aws/waiters/test_base_waiter.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/waiters/test_base_waiter.py)
> * [tests/providers/amazon/aws/triggers/test_step_function.py](https://github.com/apache/airflow/tree/main/tests/providers/amazon/aws/triggers/test_step_function.py)

Hi @Taragolis ! Thanks for your help.

I guess I was not so clear, but the corresponding code of each of that non-existing files is covered in other test files, and that is why I put those ones on the list. 

For example, one test file from the above list that does not exist is `tests/providers/amazon/aws/operators/test_emr.py`. This test file would cover the code present in `airflow/providers/amazon/aws/operators/emr.py` , right? But this code is being covered in other test files:

- `tests/providers/amazon/aws/operators/test_emr_add_steps.py`
- `tests/providers/amazon/aws/operators/test_emr_containers.py`
- `tests/providers/amazon/aws/operators/test_emr_create_job_flow.py`
- `tests/providers/amazon/aws/operators/test_emr_modify_cluster.py`
- `tests/providers/amazon/aws/operators/test_emr_notebook_execution.py`
- `tests/providers/amazon/aws/operators/test_emr_serverless.py`
- `tests/providers/amazon/aws/operators/test_terminate_job_flow.py`

So the `tests/providers/amazon/aws/operators/test_emr.py` should not be in that list.

**Comment by Taragolis** [comment]
There is pretty simple rule there if providers module placed into the `airflow/providers/amazon/aws/operators/emr.py` then tests should be placed into the `tests/providers/amazon/aws/operators/test_emr.py`.

Original check was broken for a while (1-2 years or so), so we have about 100 modules which do not follow this one.

**Comment by Lzzz666** [comment]
I’d be happy to take redis's test.

**Comment by Jasperora** [comment]
Hi, I’m new to Airflow and would like to help with Slack-related tests.

I noticed that in the current file structure, there is no `slack_notifier.py` under `providers/slack/src/airflow/providers/slack/notifications`. Only `slack_webhook.py` and `slack.py` exist.
Does it mean that there is no need for a `test_slack_notifier.py` file?

Thank you for your guidance!

**Comment by jason810496** [comment]
> Hi, I’m new to Airflow and would like to help with Slack-related tests.
> 
> I noticed that in the current file structure, there is no `slack_notifier.py` under `providers/slack/src/airflow/providers/slack/notifications`. Only `slack_webhook.py` and `slack.py` exist. Does it mean that there is no need for a `test_slack_notifier.py` file?
> 
> Thank you for your guidance!

It seems that `slack_notifier.py` was renamed to `providers/slack/src/airflow/providers/slack/notifications/slack.py`.

**Comment by jason810496** [comment]
Do we need to add unit test to cover `version_compat.py module` in each providers?
 Since I just reviewed a PR that will add test for `version_compat.py` . 
IMO, it's a special one that can be exclude from the coverage report. 

Just to align the convention as there isn't  corresponding test for `version_compat` for all current providers.

**Comment by potiuk** [comment]
> IMO, it's a special one that can be exclude from the coverage report.

Absolutely.

**Comment by dominikhei** [comment]
I'll have a look at Databricks over the weekend

**Comment by Jasperora** [comment]
I'll try helping celery's tests.

**Comment by potiuk** [comment]
@Lzzz666 We are unassigning you from this issue as part of our updated [assignment policy](https://github.com/apache/airflow/blob/main/contributing-docs/04_how_to_contribute.rst#contribute-code-changes).

This is not meant to discourage your contribution — quite the opposite! You are still very welcome to work on this issue and submit a PR for it. Simply comment that you are working on it and open a PR when ready.

We found that formal assignments were not working well, as they often prevented others from contributing when the assignee was not actively working on the issue.

**Comment by gps23** [comment]
I'd like to work on tests/providers/docker/test_exceptions.py — will submit a PR.
Update: docker/test_exceptions.py turns out to already exist — I'll work on a different missing test file instead and follow up here shortly.

**Comment by gps23** [comment]
Update: I worked on the missing HTTP provider exception test module instead and have submitted a PR for review.
PR: #69231 – Add unit tests for HTTP provider exceptions.
Looking forward to the review. Thanks!

**Comment by desusaiteja** [comment]
Hello - I picked up test_bigquery_to_sql.py (now providers/google/tests/unit/google/cloud/transfers/ after the restructure). Here is the PR: [#69788](https://github.com/apache/airflow/pull/69788)

**Comment by lbbc-lh** [comment]
Hi! I’ve opened [https://github.com/apache/airflow/pull/71844](https://github.com/apache/airflow/pull/71844) to add a dedicated test module for `BaseConfigKeys` and remove the corresponding `OVERLOOKED_TESTS` exemption as part of this issue.

**Comment by zhangxinyao88** [comment]
Hi! I’d like to work on the missing Hive plugin test module:

`providers/apache/hive/tests/unit/apache/hive/plugins/test_hive.py`

I’ll add focused coverage for `HivePlugin` and remove its entry from `OVERLOOKED_TESTS`. I’ll open a PR once it’s ready.

**Comment by zhangxinyao88** [comment]
Hi @Miretpl , I noticed this issue was closed after #69231 was merged. I still have a few related PRs open: #71970, #71983, and #71985.

When you have a chance, could you please take a look and let me know whether they still cover useful remaining work from this issue? Thank you!

**Comment by Miretpl** [comment]
Closed by GitHub when I merged the PR. Issue still relevant. I will take a look at the PRs in a free moment. Thanks!

**Comment by codewithsruthi** [comment]
I am working on the missing test module for the common AI provider exceptions:

`providers/common/ai/tests/unit/common/ai/test_exceptions.py`

That path is on the live `OVERLOOKED_TESTS` list in `test_providers_modules_should_have_tests`. I will open a PR for this slice only. Please keep this meta issue open (`related: #35442`, not closes).

---
Drafted-by: Cursor Grok 4.6 (no human review before posting)

**Comment by jroachgolf84** [comment]
Related issues (for the Contributors Workshop at Airflow Summit). When these issues are resolved, this issue will be updated to reflect that.

- https://github.com/apache/airflow/issues/72263
- https://github.com/apache/airflow/issues/72265
- https://github.com/apache/airflow/issues/72266
- https://github.com/apache/airflow/issues/72267
- https://github.com/apache/airflow/issues/72268
- https://github.com/apache/airflow/issues/72269
- https://github.com/apache/airflow/issues/72276

**Comment by keith991001** [comment]
I'm picking up the missing test module for the common.compat lineage entities:

`providers/common/compat/tests/unit/common/compat/lineage/test_entities.py`

Checked for existing claims and open PRs covering this path first — none found. Will remove the corresponding entry from `OVERLOOKED_TESTS` in `test_project_structure.py` in the same PR.

---
Drafted-by: Claude Code (Fable 5) (no human review before posting)

**Comment by keith991001** [comment]
Status note on the lineage-entities slice I picked up above: both PRs are open and awaiting first review / workflow approval —

- #72506 adds the missing test module for `common.compat.lineage.entities` (removes its `OVERLOOKED_TESTS` entry).
- #72509 fixes a real defect the test-writing surfaced: the attrs entities declare bare mutable defaults, so `Table.tags`/`columns`/`owners`/`extra` and `Column.tags` are shared class-wide and mutating one instance pollutes every other. The fix switches them to `attr.Factory`, with regression tests that fail on the previous code.

The two PRs touch the same new test file, so whichever lands second will be rebased.

---
Drafted-by: Claude Code (Fable 5) (no human review before posting)

**Comment by bingqin2** [comment]
I picked up the missing Amazon provider test module `providers/amazon/tests/unit/amazon/aws/test_exceptions.py` (for `aws/exceptions.py`). Checked for existing claims and open PRs touching that path first; none found. Writing the tests showed that `EcsTaskFailToStart` and `WaiterTerminalFailure` do not survive a pickle round-trip, so the PR fixes those too and removes the `OVERLOOKED_TESTS` entry: https://github.com/apache/airflow/pull/72824

**Comment by keith991001** [comment]
Picking up the missing test module for the Celery default configuration builder:

`providers/celery/tests/unit/celery/executors/test_default_celery.py`

Checked for existing claims and open PRs covering this path first — none found. Will remove the corresponding entry from `OVERLOOKED_TESTS` in the same PR. Scope note: `test_celery_executor.py` already covers the dict-option reload paths, so the new module focuses on the uncovered surface (`_broker_supports_visibility_timeout`, `_broker_transport_options` error paths, the SSL branch matrix, and the result-backend fallback).

---
Drafted-by: Claude Code (Fable 5) (no human review before posting)

**Comment by keith991001** [comment]
Picking up the missing test module for the common-messaging provider base class:

`providers/common/messaging/tests/unit/common/messaging/providers/test_base_provider.py`

Checked for existing claims and open PRs covering this path first — none found. Will remove the corresponding entry from `OVERLOOKED_TESTS` in the same PR.

---
Drafted-by: Claude Code (Fable 5) (no human review before posting)

**Comment by ShousenZHANG** [comment]
Taking `providers/google/tests/unit/google/cloud/links/test_bigquery_dts.py`.

It covers one class, `BigQueryDataTransferConfigLink`, and follows the three checks the sibling link tests already use — class attributes, the `persist` XCom payload, and `_format_link` against `BASE_LINK` (same shape as `test_bigquery.py` from #68066).

The `OVERLOOKED_TESTS` entry goes in the same PR.

No overlap with `test_cloud_tasks.py` (#72846). If this shape is acceptable I'd like to keep working through the other `google/cloud/links/` entries — there are 19 left after that one.

**Comment by ShousenZHANG** [comment]
Following up on my earlier comment: I have gone through the rest of the
`providers/google/tests/unit/google/cloud/links/` entries in the list. Posting the full set
here so nobody duplicates the work. Each PR adds one test module and removes only its own
`OVERLOOKED_TESTS` line.

| module | PR |
| --- | --- |
| `test_bigquery_dts.py` | #72949 |
| `test_dataprep.py` | #72955 |
| `test_datastore.py` | #72956 |
| `test_bigtable.py` | #72960 |
| `test_cloud_build.py` | #72961 |
| `test_cloud_functions.py` | #72962 |
| `test_cloud_memorystore.py` | #72963 |
| `test_cloud_sql.py` | #72964 |
| `test_compute.py` | #72965 |
| `test_data_loss_prevention.py` | #72966 |
| `test_dataflow.py` | #72967 |
| `test_dataform.py` | #72968 |
| `test_datafusion.py` | #72969 |
| `test_dataproc.py` | #72970 |
| `test_kubernetes_engine.py` | #72971 |
| `test_pubsub.py` | #72972 |
| `test_spanner.py` | #72973 |
| `test_stackdriver.py` | #72974 |
| `test_translate.py` | #72975 |
| `test_workflows.py` | #72976 |

`test_cloud_tasks.py` is deliberately left out — #72846 already covers it.

They all follow the three checks the existing `test_bigquery.py` (#68066) in the same
directory uses: the class attributes, the `persist` XCom payload, and the `_format_link`
output. A few modules needed extra cases and those are called out in the individual PRs —
`KubernetesEngineClusterLink` and the two `output_config` translate links override `persist`,
and Cloud Functions, Compute, DLP, Dataproc, Translate and the Data Fusion instance link
build their constants on `BASE_LINK` so `_format_link` does not prefix the console base a
second time.

One thing worth flagging: because these all delete adjacent lines from the same
`OVERLOOKED_TESTS` block, they will conflict with each other as they merge. I will rebase the
remainder every time one lands. If you would rather review fewer PRs, say so and I will squash
them into a handful of grouped ones instead.

**Comment by kaxil** [comment]
Closing this as obsolete.

The checklist in the description predates the provider restructure. All 182 paths in it use the old `tests/providers/...` layout, which no longer exists, so neither the checked nor the unchecked entries map onto files in the repo today. The gap itself is not closed: `OVERLOOKED_TESTS` currently carries 267 entries, more than this list ever tracked.

That in-code list is the live source of truth and `test_providers_modules_should_have_tests` enforces it, so it cannot go stale the way this description did:

https://github.com/apache/airflow/blob/d076a57dcb24303c69d8505ea4816417003ffb34/airflow-core/tests/unit/always/test_project_structure.py#L57-L63

Anyone who wants to fill one of these in can read the current list there. Open PRs already doing so are not affected by this closure.

One request for anyone picking these up: group them. A module at a time, claimed in a comment and opened as its own PR, means one review and one CI run per file, and they are all the same review. Batch a provider's worth of missing test modules into a single PR instead.
